### ISSUE TITLE
[MNG-5315] Artifact resolution sporadically fails in parallel builds

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6623]
**[Ivan Dubrov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wfrag)** opened **[MNG-5315](https://issues.apache.org/jira/browse/MNG-5315?redirect=false)** and commented

Artifact resolution can fail during the parallel build if it was downloaded during the same session.

Scenario:

1) Delete the whole Maven local repository.
2) Run build "mvn compile -T1.5C"
3) Build fails (see log extracts below)
4) If I run build again, it succeeds.

It seems like the problem is due to two thread trying to resolve same artifact concurrently. This problem never happens once I have all dependencies cached in the local repository.

Extracts from the log output:

```
Downloading: http://nexus/content/repositories/thirdparty/com/googlecode/guava-osgi/guava-osgi/11.0.0/guava-osgi-11.0.0.jar 12444/13937 KB ...
...
[DEBUG] Skipped remote update check for commons-cli:commons-cli:jar:1.2, already updated during this session.
[DEBUG] Skipped remote update check for com.googlecode.guava-osgi:guava-osgi:jar:11.0.0, already updated during this session.
[DEBUG] Skipped remote update check for org.slf4j:slf4j-api:jar:1.6.4, already updated during this session.
[DEBUG] Skipped remote update check for org.slf4j:slf4j-jdk14:jar:1.6.4, already updated during this session.
...
[ERROR] Failed to execute goal on project util: Could not resolve dependencies for project com.guidewire.pl:util:bundle:1.0-SNAPSHOT: The following artifacts could not be resolved: commons-cli:commons-cli:jar:1.2, com.googlecode.guava-osgi:guava-osgi:jar:11.0.0, org.slf4j:slf4j-api:jar:1.6.4, org.slf4j:slf4j-jdk14:jar:1.6.4: Failure to find commons-cli:commons-cli:jar:1.2 in http://nexus/content/repositories/thirdparty was cached in the local repository, resolution will not be reattempted until the update interval of gw.thirdparty has elapsed or updates are forced -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project util: Could not resolve dependencies for project com.guidewire.pl:util:bundle:1.0-SNAPSHOT: The following artifacts could not be resolved: commons-cli:commons-cli:jar:1.2, com.googlecode.guava-osgi:guava-osgi:jar:11.0.0, org.slf4j:slf4j-api:jar:1.6.4, org.slf4j:slf4j-jdk14:jar:1.6.4: Failure to find commons-cli:commons-cli:jar:1.2 in http://nexus/content/repositories/thirdparty was cached in the local repository, resolution will not be reattempted until the update interval of gw.thirdparty has elapsed or updates are forced
        at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies(LifecycleDependencyResolver.java:210)
        at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies(LifecycleDependencyResolver.java:117)
        at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved(MojoExecutor.java:258)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:201)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
        at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:167)
        at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:163)
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
        at java.util.concurrent.FutureTask.run(FutureTask.java:138)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
        at java.util.concurrent.FutureTask.run(FutureTask.java:138)
        at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
        at java.lang.Thread.run(Thread.java:662)
Caused by: org.apache.maven.project.DependencyResolutionException: Could not resolve dependencies for project com.guidewire.pl:util:bundle:1.0-SNAPSHOT: The following artifacts could not be resolved: commons-cli:commons-cli:jar:1.2, com.googlecode.guava-osgi:guava-osgi:jar:11.0.0, org.slf4j:slf4j-api:jar:1.6.4, org.slf4j:slf4j-jdk14:jar:1.6.4: Failure to find commons-cli:commons-cli:jar:1.2 in http://nexus/content/repositories/thirdparty was cached in the local repository, resolution will not be reattempted until the update interval of gw.thirdparty has elapsed or updates are forced
        at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve(DefaultProjectDependenciesResolver.java:189)
        at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies(LifecycleDependencyResolver.java:185)
        ... 16 more
Caused by: org.sonatype.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: commons-cli:commons-cli:jar:1.2, com.googlecode.guava-osgi:guava-osgi:jar:11.0.0, org.slf4j:slf4j-api:jar:1.6.4, org.slf4j:slf4j-jdk14:jar:1.6.4: Failure to find commons-cli:commons-cli:jar:1.2 in http://nexus/content/repositories/thirdparty was cached in the local repository, resolution will not be reattempted until the update interval of gw.thirdparty has elapsed or updates are forced
        at org.sonatype.aether.impl.internal.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:375)
        at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve(DefaultProjectDependenciesResolver.java:183)
        ... 17 more
Caused by: org.sonatype.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: commons-cli:commons-cli:jar:1.2, com.googlecode.guava-osgi:guava-osgi:jar:11.0.0, org.slf4j:slf4j-api:jar:1.6.4, org.slf4j:slf4j-jdk14:jar:1.6.4: Failure to find commons-cli:commons-cli:jar:1.2 in http://nexus/content/repositories/thirdparty was cached in the local repository, resolution will not be reattempted until the update interval of gw.thirdparty has elapsed or updates are forced
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:538)
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:216)
        at org.sonatype.aether.impl.internal.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:358)
        ... 18 more
Caused by: org.sonatype.aether.transfer.ArtifactNotFoundException: Failure to find commons-cli:commons-cli:jar:1.2 in http://nexus/content/repositories/thirdparty was cached in the local repository, resolution will not be reattempted until the update interval of gw.thirdparty has elapsed or updates are forced
        at org.sonatype.aether.impl.internal.DefaultUpdateCheckManager.newException(DefaultUpdateCheckManager.java:230)
        at org.sonatype.aether.impl.internal.DefaultUpdateCheckManager.checkArtifact(DefaultUpdateCheckManager.java:176)
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:427)
        ... 20 more

```

---

**Affects:** 3.0.4

**Attachments:**
- [MNG-5315-3.0.x.patch](https://issues.apache.org/jira/secure/attachment/12713236/MNG-5315-3.0.x.patch) (_11.24 kB_)
- [MNG-5315-3.1.2-SNAPSHOT.patch](https://issues.apache.org/jira/secure/attachment/12713390/MNG-5315-3.1.2-SNAPSHOT.patch) (_12.14 kB_)

**Issue Links:**
- [MNG-2802](https://issues.apache.org/jira/browse/MNG-2802) Concurrent-safe access to local Maven repository
 (_**"is a child of"**_)
- [MRESOLVER-114](https://issues.apache.org/jira/browse/MRESOLVER-114) ArtifactNotFoundExceptions when building in parallel
 (_**"is caused by"**_)
- [MNG-4706](https://issues.apache.org/jira/browse/MNG-4706) Multithreaded building can create bad files for downloaded artifacts in local repository

- [MNG-6996](https://issues.apache.org/jira/browse/MNG-6996) Upgrade Maven Resolver to 1.6.1
 (_**"requires"**_)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

22 votes, 35 watchers

**Comment by jira-importer** [comment]
**[Christoph Kutzinski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kutzi@gmx.de)** commented

We're seeing similar errors in our CI environment (Jenkins) quite frequently - i.e. several times a day.

This causes major manual effort to determine the cause of the failure, restart the build and wait

**Comment by jira-importer** [comment]
**[Sandy McPherson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sandymcp)** commented

We're also seeing this using Teamcity and Nexus. Which repository are you guys using? Could the root problem be Nexus related? Would appear that maven does not find the dependency in Nexus. We have a couple of Nexus repositories which are being searched.

**Comment by jira-importer** [comment]
**[Ivan Dubrov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wfrag)** commented

No, it's not repository related. The issue never happens in a single-threaded build. It looks like it happens due to the race-condition between two threads trying to download the same artifact at the same time. Also, even though build is failed the dependency is actually downloaded into the local repository (so you can see it there after the build fails).

Something like (just speculating):

1) first thread looks into the local repository, nothing there, goes to the remote repository
2) first thread downloads remote dependency, but does not update _maven.repositories yet.
3) second thread looks into the local repository, sees the dependency, but _maven.repositories does not contain remote repository id!
4) second thread fails with "Could not resolve dependencies", since it thinks dependency is not available from the remote repository.
5) first thread updates _maven.repositories. Next time build is invoked, dependency is used from the local repository.

**Comment by jira-importer** [comment]
**[Sandy McPherson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sandymcp)** commented

After a few false starts I found a plausible solution.

I Patched to only allow one thread at a time to do dependency resolution. I tried to do this with finer granularity (per artifact) but failed miserably due to the program structure. However the performance gain we have achieved in the builds we were targeting was one order of magnitude!

```
DefaultProjectDependenciesResolver

    public DependencyResolutionResult resolve(
            DependencyResolutionRequest request)
            throws DependencyResolutionException {
        long start = System.currentTimeMillis();
        System.out.println("0 **** locking "+DefaultProjectDependenciesResolver.class);
        synchronized (DefaultProjectDependenciesResolver.class) {
            System.out.println((System.currentTimeMillis()-start)+" **** locked "+DefaultProjectDependenciesResolver.class);
            RequestTrace trace = RequestTrace.newChild(null, request);
.
.
.
.
.
            System.out.println((System.currentTimeMillis()-start)+" **** unlocked "+DefaultProjectDependenciesResolver.class);
            return result;
        }
```

Results are now

No failures so far. Massive speed improvements on larger builds with similar dependencies:

serial: 10-11m
-T 1C  1.5m
-T 2C 1.25m
-T 3C nearly 1m

**Comment by jira-importer** [comment]
**[Christoph Kutzinski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kutzi@gmx.de)** commented

Sandy, this sounds great - when will it be available in a release?

BTW: did I understand you correctly: you made it single-threaded and it became faster?

**Comment by jira-importer** [comment]
**[Sandy McPherson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sandymcp)** commented

Making the parallel builds work gave us the performance increase. Only dependency resolution is restricted to one thread at a time, due to some race condition, which I didn't attempt to analyse in depth.

I don't know when it will be in a release, I am not part of the maven developers group. It is however straightforward to patch maven by downloading the source and adding the synchronize where I did.

**Comment by jira-importer** [comment]
**[Karthikeyan Kulasekaran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karthik85_4u)** commented

Please fix this issue as well MNG-5252. This is related to MNG-5315

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Can you please re-test with maven 3.1.0 ? The root cause of this issue may be fixed there.

**Comment by jira-importer** [comment]
**[Sandy McPherson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sandymcp)** commented

My first fix was with a 3.1-SNAPSHOT from the beginning of June. I then went back and patched 3.0.5 as we didn't want to use a snapshot. The 3.1-SNAPSHOT I checked out a version from June 2 21:32:45 (last committer Arnaud Heritier). This was definitely still broken.

**Comment by jira-importer** [comment]
**[Sami Salonen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=salsa)** commented

Thanks a lot Sandy for your solution.

We used to suffer from this very same problem on a daily basis. I applied your patch to both maven 3.0.5 and 3.1.0. Now, after one month and about 3000 builds later, not a single build has failed for this problem anymore. Perfect.

**Comment by jira-importer** [comment]
**[Stephen Evanchik](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=evanchsa)** commented

Simple synchronization as proposed in this issue against 3.1.2-SNAPSHOT

**Comment by jira-importer** [comment]
**[Stephen Evanchik](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=evanchsa)** commented

Patch suitable for the 3.0.x branch

**Comment by jira-importer** [comment]
**[Stephen Evanchik](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=evanchsa)** commented

Alternatively one can create a custom Component that adds this without modifying maven-core. Here is an example against 3.1.2-SNAPSHOT and a 3.0.x:

https://github.com/evanchsa/maven-parallel-resolver.git

**Comment by jira-importer** [comment]
**[Mahout QA](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mahoutqa)** commented

https://builds.apache.org/job/Hama-trunk/274/console

Is this the same issue?

---

Oh, sorry. I've changed version of rat plugin.

**Comment by jira-importer** [comment]
**[Esteban Porcelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=windchess)** commented

This issue is marked with priority: minor

This is not a minor issue. We are using Maven in a large enterprise project.
The project takes one third the time to build using parallel builds (which is the reason for having parallel builds).
But intermittent failures every other day are not acceptable as it requires manual intervention.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

If you are on a reasonably newer version of Maven you can try: http://takari.io/book/30-team-maven.html#concurrent-safe-local-repository

Or you can now use it as a core extension with Maven 3.3.1: http://takari.io/2015/03/19/core-extensions.html

**Comment by jira-importer** [comment]
**[Carlos Alexandro Becker](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=caarlos0)** commented

Happens here too... very annoying BTW

**Comment by jira-importer** [comment]
**[Jorg Heymans](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=heymjo)** commented

I added the takari libs into our maven 3.2.5 install as suggested above but it did not fix the problem. I have a build that consistently fails with -T1C option so it's easy to test any other solution you may have.

**Comment by jira-importer** [comment]
**[Kogler Hannes (HKO 4530)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hannes.kogler@ntsretail.com)** commented

Dear Sir or Madam,

Thank you for your message. I am out of the office at the moment and will be back on Monday, 17th. August 2015.
I will reply to your e-mail as soon as possible when I return.

Please note that I will have limited access to my e-mail and your message will not be forwarded automatically.

For urgent issues may I refer you to my colleagues Michael Weichselbaumer (MWE) and Ronald Angerer (RAN).

Best regards

Hannes Kogler
Software Engineer

Registered Office | NTS New Technology Systems GmbH
Doernbacherstraße 126 | 4073 Wilhering | Austria
FN155693k | ATU42651803 | Court of Jurisdiction Linz

**Comment by jira-importer** [comment]
**[Stephen Evanchik](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=evanchsa)** commented

I'm not sure that the Takari libs work for Maven 3.2.x.

The Takari libs + Maven 3.3.x are an improvement but parallel builds will still fail. This is due to certain projects having their classpath corrupted. Successful builds have the correct classpath (obviously); unsuccessful builds will have an empty classpath.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

If you can reproduce it in a test project I can do something about it. The concurrent connector was originally created for a customer with 800 jobs in Jenkins sharing the same local repository and for this case it worked. I don't have any other solutions and I can't guess and make things for you to try. It could be the plugins you're using or some other configuration but if I, myself, cannot reproduce it there's nothing I can do to help you.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

I have some work that involves concurrent/distributed builds so I'm likely to take a look in the coming months but I'm happy to look sooner if you can make a build that exhibits the issue. Or if you look at the test cases for the connector and can make something fail I'll have something to work with. That's not a guarantee I have time to fix it but that's the minimum requirement for me to even try to do something about it.

**Comment by jira-importer** [comment]
**[Paul Polishchuk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lngtmg)** commented

really annoying thing :(

**Comment by jira-importer** [comment]
**[Michel Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=micheljung)** commented

Yes, we run our release build on an empty local repository, and this issue prevents us from running parallel builds as they fail every now and then.

**Comment by jira-importer** [comment]
**[Bob Tiernay](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=btiernay)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) Is this still an open issue or was this fixed as a consequence of [MNG-6110](https://issues.apache.org/jira/browse/MNG-6110) and can now be closed?

**Comment by jira-importer** [comment]
**[Alexey Yudichev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ayud@mail.ru)** commented

A build using 3.5.0 did not fail for me, while 3.2 had always been failing when the local repo was clean. Windows 7, 12 threads, 8 CPU cores.

**Comment by jira-importer** [comment]
**[Debraj Manna](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuk)** commented

3.5.0 is working fine for me for the past 1 year.

**Comment by jira-importer** [comment]
**[Sami Salonen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=salsa)** commented

I can confirm that the recent Maven versions have worked fine for me also. Perhaps it was 3.3.9 that was stable already.

**Comment by jira-importer** [comment]
**[Alexey Yudichev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ayud@mail.ru)** commented

Just 2 weeks ago I saw 3.3.9 concurrent build downloading same arrtifact (dependency POM) concurrently and corrupting it in the local repo as a result. Have not seen this on some of the more recent versions.

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

Close because it is no longer seen by users with the latest Maven versions.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Read my description on the issue: https://issues.apache.org/jira/browse/MRESOLVER-114?focusedCommentId=17127769&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17127769

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Please test a new, global locking approach with MRESOLVER-123. It is suited for concurrent access in **one** VM.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

MRESOLVER-131 introduces a fully new, multi-JVM concurrent approach based on Redis. Try it from here: https://github.com/apache/maven-resolver/tree/MRESOLVER-131. Read it from here: https://maven.apache.org/resolver-archives/resolver-LATEST/maven-resolver-synccontext-redisson/index.html

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Use any of the two new sync contexts to solve this concurrency issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Maven Resolver 1.6.2 will be available with Maven 3.8.2 so this has been addressed in 3.8.x too.
